### PR TITLE
Add exclusion of incomplete records

### PR DESCRIPTION
**Pull request by brinxmat**
Excludes documents that would not be searchable (but are likely never sent for processing)
Fixes a number of issues, test coverage, URI management and JSON generation

### CODE REVIEW

File: search-resources-api/src/main/java/no/unit/nva/dynamodb/DynamoDBStreamHandler.java [hunk 1]
```diff
@@ -112,7 +115,12 @@ private boolean isNotValidEventName(String eventName) {
     private void upsertSearchIndex(DynamodbEvent.DynamodbStreamRecord streamRecord) throws SearchException {
         logStreamRecord(streamRecord);
         DynamoDBEventTransformer eventTransformer = new DynamoDBEventTransformer();
-        IndexDocument document = eventTransformer.parseStreamRecord(streamRecord);
+        IndexDocument document = Optional.ofNullable(eventTransformer.parseStreamRecord(streamRecord)).stream()
+                .filter(Objects::nonNull)
+                .collect(SingletonCollector.collectOrElse(null));
+        if (isNull(document)) {
```
Comment (axthosarouris) [inline search-resources-api/src/main/java/no/unit/nva/dynamodb/DynamoDBStreamHandler.java:121]: cleaner if we write 

```
 private void upsertSearchIndex(DynamodbEvent.DynamodbStreamRecord streamRecord) throws SearchException {
        logStreamRecord(streamRecord);
        DynamoDBEventTransformer eventTransformer = new DynamoDBEventTransformer();
        var document = eventTransformer.parseStreamRecord(streamRecord);
        if (nonNull(document)) {
            elasticSearchClient.addDocumentToIndex(document);
        }
    }
```

File: search-resources-api/src/main/java/no/unit/nva/search/IndexDate.java [hunk 2]
```diff
@@ -20,6 +22,12 @@
     private final String month;
     private final String day;
 
+    /**
+     * Constructor for basic serialization of IndexDate.
```
Comment (axthosarouris) [inline search-resources-api/src/main/java/no/unit/nva/search/IndexDate.java:26]: Constructor for basic **de**serilization ...

File: search-resources-api/src/test/java/no/unit/nva/dynamodb/DynamoDBStreamHandlerTest.java [hunk 3]
```diff
@@ -186,11 +210,29 @@ public void testHandleExceptionInEventHandlingShouldGiveException() throws IOExc
...
+    public void dynamoDBStreamHandlerCreatesHttpRequestWithIndexDocumentWithModifyEventValidRecord()
+            throws IOException {
+        URI identifier = URI.create("https://example.org/publication/1006a");
+        String contributorIdentifier = "123";
+        String contributorName = "Bólsön Kölàdỳ";
+        List<Contributor> contributors = Collections.singletonList(
+                generateContributor(contributorIdentifier, contributorName, 1));
+        String mainTitle = "Moi buki";
+        String type = "Book";
+        IndexDate date = new IndexDate("2020", "09", "08");
```
Comment (axthosarouris) [inline search-resources-api/src/test/java/no/unit/nva/dynamodb/DynamoDBStreamHandlerTest.java:224]: Refactor to a method with name bookEntryWithNoEmptyFields or sth like this.

File: search-resources-api/src/test/java/no/unit/nva/dynamodb/DynamoDBStreamHandlerTest.java [hunk 4]
```diff
@@ -53,321 +55,312 @@
...
         assertThat(exception.getMessage(), containsString(ENVIRONMENT_VARIABLE_NOT_SET));
     }
 
     @Test
-    @DisplayName("testHandlerHandleSimpleRemoveEventWithoutProblem")
-    public void testHandlerHandleSimpleRemoveEventWithoutProblem() throws IOException {
-        DynamodbEvent requestEvent = loadEventFromResourceFile(SAMPLE_REMOVE_EVENT_FILENAME);
-        String response = handler.handleRequest(requestEvent, context);
-        assertNotNull(response);
+    void handlerReturnsSuccessMessageWhenDocumentDoesNotExist() throws IOException {
```
Comment (axthosarouris) [inline search-resources-api/src/test/java/no/unit/nva/dynamodb/DynamoDBStreamHandlerTest.java:110]: I think we should check that the ElasticSearch client is being called. You can use the Mockito expression verify for that

```
verify(restClient, (atLeastOnce()));
verify(restClient, (atMostOnce()));
verify(restClient, (times(1)));  
```

File: search-resources-api/src/test/java/no/unit/nva/dynamodb/DynamoDBStreamHandlerTest.java [hunk 5]
```diff
@@ -53,321 +55,312 @@
...
+        assertThat(cause.getMessage(), containsString(DynamoDBStreamHandler.EMPTY_EVENT_NAME_ERROR));
 
+        assertThat(testAppender.getMessages(), containsString(DynamoDBStreamHandler.LOG_MESSAGE_MISSING_EVENT_NAME));
+    }
 
-        InputException cause = (InputException) exception.getCause();
+    @ParameterizedTest
+    @DisplayName("handler returns success message when event type is {0}")
+    @ValueSource(strings = {INSERT, MODIFY, REMOVE})
+    void handlerReturnsSuccessWhenEventNameIsValid(String eventName) throws IOException {
```
Comment (axthosarouris) [inline search-resources-api/src/test/java/no/unit/nva/dynamodb/DynamoDBStreamHandlerTest.java:153]: ditto

File: search-resources-api/src/test/java/no/unit/nva/search/SearchResourcesApiHandlerTest.java [hunk 6]
```diff
@@ -43,12 +44,11 @@ public void defaultConstructorThrowsIllegalStateExceptionWhenEnvironmentNotDefin
...
     @Test
-    public void processInputReturnsNullWhenInputIsEmpty() throws ApiGatewayException {
+    public void processInputReturnsNullWhenInputIsEmpty() {
         RequestInfo requestInfo = mock(RequestInfo.class);
         Context context = mock(Context.class);
-        assertThrows(ApiGatewayException.class, () ->  searchResourcesApiHandler.processInput(null,
-                requestInfo,
-                context));
+        Executable executable = () -> searchResourcesApiHandler.processInput(null, requestInfo, context);
+        assertThrows(ApiGatewayException.class, executable);
```
Comment (axthosarouris) [inline search-resources-api/src/test/java/no/unit/nva/search/SearchResourcesApiHandlerTest.java:51]: Very generic test that can succeed for other reasons than that expected here. SearchException is thrown at the moment. Consider deleting the test if this case is covered elsewhere

**Review by axthosarouris — CHANGES_REQUESTED** [review]

**Review by svenngjeto — DISMISSED** [review]

**Review by svenngjeto — DISMISSED** [review]

**Review by axthosarouris — DISMISSED** [review]

File: search-resources-api/src/main/java/no/unit/nva/search/IndexDocument.java [hunk 7]
```diff
@@ -92,9 +105,15 @@ public int hashCode() {
         return Objects.hash(type, id, contributors, title, date);
     }
 
+    private String getFinalPathElementFromUri() {
+        int lastPathElementSeparator = id.toString().lastIndexOf(PATH_SEPARATOR) + 1;
```
[the remark this answers is not in the document]
Comment (brinxmat) [reply search-resources-api/src/main/java/no/unit/nva/search/IndexDocument.java]: Added checking, we should fix this issue across the board in commons

**Review by axthosarouris — APPROVED** [review]

**Pull request merged**